### ISSUE TITLE
Character encodings may not be respected

### ISSUE THREAD

**Issue by gingerbeardman** [thejoelpatrol/fusehfs#4]
Following on from thread.

Previously (hmm, when?) I could set my Mac to Japanese and reboot, mount an HFS disc that uses MacJapanese character encoding and see the filenames as intended. 

Such foreign discs are tricky as they contain filenames in multiple character sets. Files may have been copied from other discs or downloaded from the internet. 

I have found that Tcl has good encodings, most of them written by Apple back in the mid-1990s when this stuff was still very much current. 

My own tools for processing these discs use hfsutils to get the raw filenames and then I convert them using a Tcl one-liner. 

Another gotcha is that bugs/omissions in Japanese input methods (apps/accessories that assist typing of complex script/alphabets like Japanese) allowed non-displayable characters to be typed in filenames! Example: when renaming a file pressing delete on an Extended keyboard would insert that invisible character rather than deleting anything. 

If you need more details please ask as I have the info in my notes that I can dig out. 

## Sample HFS images 

disk images that have a mix of MacRoman and MacJapanese:
- https://archive.org/details/nikkei-mac-cd-rom-1994-2-17
- https://archive.org/details/mac-treasure-chest-best-games-101
- https://archive.org/details/cd-rom-maclife-09 

This one looks 99% MacRoman, with a minimal file or two with MacJapanese names:
- https://archive.org/details/info-mac-may-1993-japanese

**Comment by thejoelpatrol** [comment]
I'm not sure macOS supports MacJapanese any more. The closest I can find with `$ iconv -l` is `SHIFT_JISX0213`. Forcing that encoding for the whole volume results in this:
<img width="858" alt="image" src="https://user-images.githubusercontent.com/5752950/188236399-ab1f48e6-2f39-4623-812e-e8c3f7aaa16c.png">
AFAICT from Google Translate these appear to be roughly sensible filenames. 

Of course we don't want to force this manually globally. But even forcing one encoding that we decide via heuristic or settings/preferences or whatever would be better than assuming everything is MacRoman, as we are doing now in the absence of a command-line flag. We could do something like you mention, getting the system language and choosing the appropriate classic Mac encoding based on that, e.g. setting MacJapanese if you set your language to Japanese, MacHebrew if you are set to Hebrew, or Mac OS Thai, Mac OS Ukrainian, etc. But not all of these contain ASCII in the first 128 bits, so you won't be able to make any sense of MacRoman disks if you are set to Thai or Ukrainian, for example, so that's not great. 

I wonder if a simple GUI application would be useful here, to drag and drop a disk image on it and you can set the language/encoding there for people who don't want to use the command line.

I don't know what kind of heuristic could detect files with multiple encodings per volume, though. That seems real tough. Would those have been handled correctly by classic Mac OS?

In the interim, if you would like to use a volume that you know has a particular encoding, you can mount it manually. Open it in Disk Utility, and unmount the volume but do not eject the image. Get the disk number, e.g. `/dev/disk2s2`. Then, run this:

`$  /Library/Filesystems/fusefs_hfs.fs/Contents/Resources/mount_fusefs_hfs --encoding=${ENCODING_NAME} ${DISK_NUMBER} ${MOUNTPOINT}` 
eg:
`$  /Library/Filesystems/fusefs_hfs.fs/Contents/Resources/mount_fusefs_hfs --encoding=SHIFT_JISX0213 /dev/disk2s2 /Users/joel/mnt`

**Comment by gingerbeardman** [comment]
MacJapanese is closely related to SHIFT-JIS but they're not the same, [see here](https://en.wikipedia.org/wiki/Shift_JIS#MacJapanese). That said, it may be close enough to be a workable solution.

I think your suggestions are all good. If the encoding could be remembered between mounts that would  be great. 

Or maybe the user could add something to the filename that would clue fusehfs in to the required encoding? That way nothing would need to be stored on disk.

I'll try manual mounting soon.

**Comment by d235j** [comment]
It looks like iconv doesn't support MacJapanese but CoreFoundation does. The downside of rewriting the character encoding code using CF is that it would make fusehfs less portable to Linux.

I'm guessing something is stored on disk indicating encoding. Will need to investigate.

**Comment by gingerbeardman** [comment]
> I'm guessing something is stored on disk indicating encoding. Will need to investigate.

I don't believe there is, but I don't have any references to cite. The language of the host OS is responsible for interpreting the filenames according to its default encoding. Very old school.

That said, I would be interested to see what you find!

**Comment by joevt** [comment]
There's a text encoding hint in the Finder Info of the Master Director Block. Maybe it's related? See `dumpencoding` at:
 https://gist.github.com/joevt/a99e3af71343d8242e0078ab4af39b6c
See `GET_HFS_TEXT_ENCODING` at:
 https://github.com/apple-oss-distributions/hfs/blob/hfs-627.40.1/mount_hfs/mount_hfs.c

There's HFS Encoding kexts for various macOS versions from Sierra to Catalina at:
`/System/Library/Filesystems/hfs.fs/Contents/Resources/Encodings/`
HFS_MacArabic.kext
HFS_MacCentralEurRoman.kext
HFS_MacChineseSimp.kext
HFS_MacChineseTrad.kext
HFS_MacCroatian.kext
HFS_MacCyrillic.kext
HFS_MacGreek.kext
HFS_MacHebrew.kext
HFS_MacIcelandic.kext
HFS_MacJapanese.kext
HFS_MacKorean.kext
HFS_MacRomanian.kext
HFS_MacThai.kext
HFS_MacTurkish.kext

I think there might be source code at least for converting MacJapanese?
 https://github.com/apple-oss-distributions/hfs/blob/main/hfs_japanese/hfs_japanese.kmodproj/JapaneseConverter.c
And also MacRoman:
 https://github.com/apple-oss-distributions/hfs/blob/main/hfs_encodings/hfs_encodings.c

The documentation archive has a note about an Encoding popup in the Finder's Get Info window for an HFS Standard volume:
 https://developer.apple.com/library/archive/qa/qa1173/_index.html#//apple_ref/doc/uid/DTS10001705
What does that popup look like? Is it changing the value in the Master Director Block?

More about text encodings for HFS Plus:
 https://developer.apple.com/library/archive/technotes/tn/tn1150.html#//apple_ref/doc/uid/DTS10002989
It implies that HFS Standard does not have the per-file or per-folder text encoding settings that HFS Plus has and that the text encoding "varies depending on how the system software was localized and what language kits are installed".

**Comment by d235j** [comment]
Looks like if it exists at all, it’s stored in the Finder info word in the MDB?
https://github.com/apple-oss-distributions/hfs/blob/4e3719273a0c670ef4aa7c77bb421c89f3473e14/mount_hfs/mount_hfs.c

**Comment by gingerbeardman** [comment]
Interesting!

Tcl has good conversion routines and encoding tables which were written by Apple themselves.

**Comment by thejoelpatrol** [comment]
> Looks like if it exists at all, it’s stored in the Finder info word in the MDB? https://github.com/apple-oss-distributions/hfs/blob/4e3719273a0c670ef4aa7c77bb421c89f3473e14/mount_hfs/mount_hfs.c

good find @d235j, I'll look into this

**Comment by joevt** [comment]
Here's another Japanese one from infinite-mac:
 https://github.com/mihaip/infinite-mac/blob/main/Images/KanjiTalk%207.5.3%20HD.dsk
You may need to install Git LFS (large file support) to get this image.

The macOS Anthology 99 Disk 1 has HFS images for other languages. These are Disk Copy 4.2 disk images which you can easily convert to normal disk images using `dd`.
 https://macintoshgarden.org/apps/mac-os-anthology-99

**Comment by joevt** [comment]
None of the Disk Copy 4.2 disk images on "Mac OS Anthology '99 Disk 1" appear to have a text encoding set in the MDB.

For disk images with non Mac-Roman names, you can try using `dd` and `xxd` to extract the volume name from the mdb, and `iconv` to convert to utf-8.

```
cd "/Volumes/Mac OS Anthology '99 Disk 1"

# Search for disk images that have a text encoding
IFS=$'\n'
for thefile in $(find . -type f); do
	if [[ "BD" == $(dd if="$thefile" bs=1 count=2 skip=0x454 2> /dev/null) ]]; then
		if [[ "00000000" != $(dd if="$thefile" bs=1 count=4 skip=$((0x454+0x6C)) 2> /dev/null | xxd -p) ]]; then
			echo "$thefile"
		fi
	fi
done

# Get all names
IFS=$'\n'
for thefile in $(find . -type f); do
	if [[ "BD" == $(dd if="$thefile" bs=1 count=2 skip=0x454 2> /dev/null) ]]; then
		namelen=$((0x$(dd if="$thefile" bs=1 count=1 skip=$((0x454+0x24)) 2> /dev/null | xxd -p) ))
		dd if="$thefile" of=/tmp/volname.txt bs=1 count="$namelen" skip=$((0x454+0x25)) 2> /dev/null
		printf "%-100s" "$thefile"
		iconv -f MACROMAN -t UTF8 /tmp/volname.txt 2>&1
		printf "\n"
	fi
done

# Get all Arabic names
IFS=$'\n'
for thefile in $(find . -type f -path '*Arabic*'); do
	if [[ "BD" == $(dd if="$thefile" bs=1 count=2 skip=0x454 2> /dev/null) ]]; then
		namelen=$((0x$(dd if="$thefile" bs=1 count=1 skip=$((0x454+0x24)) 2> /dev/null | xxd -p) ))
		dd if="$thefile" of=/tmp/volname.txt bs=1 count="$namelen" skip=$((0x454+0x25)) 2> /dev/null
		iconv -f MACARABIC -t UTF8 /tmp/volname.txt 2>&1
		printf "\n"
	fi
done

# Get all Hebrew names
IFS=$'\n'
for thefile in $(find . -type f -path '*Hebrew*'); do
	if [[ "BD" == $(dd if="$thefile" bs=1 count=2 skip=0x454 2> /dev/null) ]]; then
		namelen=$((0x$(dd if="$thefile" bs=1 count=1 skip=$((0x454+0x24)) 2> /dev/null | xxd -p) ))
		dd if="$thefile" of=/tmp/volname.txt bs=1 count="$namelen" skip=$((0x454+0x25)) 2> /dev/null
		iconv -f MACHEBREW -t UTF8 /tmp/volname.txt 2>&1
		printf "\n"
	fi
done
```

**Comment by gingerbeardman** [comment]
> let's find some other languages.

You can find some Eastern European and more at offender archive. Perhaps the easiest way to find them is using DiscMaster 
https://discmaster.textfiles.com

**Comment by thejoelpatrol** [comment]
I agree that the Finder Info field in the MDB doesn't contain these hints for the disks I've checked, so that probably won't help often enough to really solve this. I wonder if that field started to be populated by later version of Mac OS once HFS+ came on the scene and it was already recognized that character encodings had become a problem. 

So, are we back to the idea of a helper app that simplifies manual mounting with a desired text encoding? I don't particularly love the idea of using some convention in the disk image file name to indicate the encoding, because it's non-obvious and would restrict what you could name image files. 

I'm going to ignore the possibility that volumes could contain files with names in multiple character sets if we don't have reason to believe that classic Mac OS ever would have been able to handle that.

**Comment by joevt** [comment]
> The documentation archive has a note about an Encoding popup in the Finder's Get Info window for an HFS Standard volume:
> https://developer.apple.com/library/archive/qa/qa1173/_index.html#//apple_ref/doc/uid/DTS10001705
> What does that popup look like? Is it changing the value in the Master Director Block?

I tested this in Leopard.
<img width="637" alt="Image" src="https://github.com/user-attachments/assets/c7fe3d4f-8235-45a2-9361-a033016ac95e" />

The disk is mounted as Roman by default. Changing the Encoding option to Arabic unmounts and remounts the HFS file system. The selecting encoding is not saved to the MDB or anywhere else.

I can set the encoding in the MDB of the disk image file like this (04 is the MacArabic encoding):
```
printf "enc\x04" | dd of="/Volumes/Shared Folders/Home/AB-Arabic/ À»Í ≠±.dmg" bs=1 oseek=$((512*2+0x6c)) conv=notrunc
```
With this change, Leopard will mount the disk image with the correct encoding already selected.

How to detect the encoding:
I suppose you could loop through all the directory and file names and try to decode each one with all of the different encodings and use statistics to determine the best one (assuming the correct encoding will have less illegal encodings than others). However, that won't work if an incorrect encoding doesn't have illegal encodings... In that case, you might need to use a dictionary to check the words and that is probably way beyond the scope of this project.

**Comment by gingerbeardman** [comment]
> I suppose you could loop through all the directory and file names and try to decode each one with all of the different encodings and use statistics to determine the best one (assuming the correct encoding will have less illegal encodings than others). However, that won't work if an incorrect encoding doesn't have illegal encodings... In that case, you might need to use a dictionary to check the words and that is probably way beyond the scope of this project.

This is how The Unarchiver does it when decompressing vintage Macintosh files, but it's not 100% reliable so they present a similar encoding picker dropdown when they can't be certain.

**Comment by thejoelpatrol** [comment]
> I suppose you could loop through all the directory and file names and try to decode each one with all of the different encodings and use statistics to determine the best one (assuming the correct encoding will have less illegal encodings than others). However, that won't work if an incorrect encoding doesn't have illegal encodings... In that case, you might need to use a dictionary to check the words and that is probably way beyond the scope of this project.

I thought about something like this, but it's still a heuristic and also a lot of work, so I agree that's not how I want to do it.

> I can set the encoding in the MDB of the disk image file like this (04 is the MacArabic encoding):

This is not a bad idea. The hypothetical helper tool for mounting could optionally set the hint in the volume, so you only have to do it once, assuming we start reading that in fuseHFS. Or the helper could even just make setting the encoding easier than manually mucking around with the bytes.

**Comment by d235j** [comment]
The old equates included with MPW and some additional files around indicate that the script is stored as a flag in the FXInfo and DXInfo structs, in what fusehfs currently has there as unused.

**Comment by joevt** [comment]
HFS is 1985.
HFS+ is 1998.
`fdScript` and `frScript` is at most 1990. Not sure about earlier?
They have a validity bit so you can test if the script code should be used or not.

**Comment by joevt** [comment]
I have not tested changing the `fdScript` and `frScript` but we can do it using the `com.apple.FinderInfo` xattr.

xattr `com.apple.FinderInfo` layout:

```
; FInfo (Finder Information) record layout

$00 OSType fdType;          ; type of file [long]
$04 OSType fdCreator;       ; file's creator [long]
$08 unsigned short fdFlags; ; flags [word]
$0A Point fdLocation;       ; file's location [point]
$0E short fdFldr;           ; file's window [word]

; FXInfo record layout

$10 short fdIconID;         ; Icon ID [word]
$12 short fdUnused[3];      ; unused but reserved [3 words]
$18 char fdScript;          ; script flag and number [byte]
$19 char fdXFlags;          ; more flags bits [byte]
$1A short fdComment;        ; Comment ID [word]
$1C long fdPutAway;         ; Home Dir ID [2 words]

; DInfo record layout for directories

$00 Rect frRect;            ; Folder Rect [4 words]
$08 unsigned short frFlags; ; Flags [word]
$0A Point frLocation;       ; Location [2 words]
$0E short frView;           ; Folder view [word]

; DXInfo record layout

$10 Point frScroll;         ; scroll position [2 words]
$14 long frOpenChain;       ; dirID chain of open folders [2 words]
$18 char frScript;          ; script flag and number [byte]
$19 char frXFlags;          ; more flags bits [byte]
$1A short frComment;        ; comment [word]
$1C long frPutAway;         ; Dir ID [2 words]
```

We see that `fdScript` and `frScript` are both at offset 0x18 in the `com.apple.FinderInfo` xattr.

In Leopard for the arabic `Install-1/System Plugins/Arabic System Plugin` file, here's the xattr for each of those folders and file.
```
xattr -p com.apple.FinderInfo "/Volumes/تثبيت-‮١‬" | xxd | sed -E '/^0000020: 0a +\.$/d'
0000000: 002e 0009 00df 013f 0100 005c 0248 0100  .......?...\.H..
0000010: ffee ffde 0000 0000 0000 373d 0000 0000  ..........7=....

xattr -p com.apple.FinderInfo "/Volumes/تثبيت-‮١‬/ملحقات النظام" | xxd | sed -E '/^0000020: 0a +\.$/d'
0000000: 0043 000b 00b1 00e9 0100 0040 00c0 0100  .C.........@....
0000010: fff8 ffe5 0000 0000 0000 8000 0000 0000  ................

xattr -p com.apple.FinderInfo "/Volumes/تثبيت-‮١‬/ملحقات النظام/ملحق النظام العربي" | xxd | sed -E '/^0000020: 0a +\.$/d'
0000000: 7363 7269 4142 7379 3100 0000 0040 0000  scriABsy1....@..
0000010: 1e69 0000 0000 0000 0000 0000 0000 0000  .i..............
```

We see that `fdScript` and `frScript` are not valid.

Maybe they get set using the MDB script when copying the file to a different HFS of HFS+ disk? HFS+ doesn't need the script except maybe as a hint for copying to an HFS disk.

I wonder if the script for the folder is just for the folder name, or if it affects the names of children or grandchildren that don't have a script?

There's only a `com.apple.FinderInfo` for the root directory but the script there doesn't match the script in the MDB.
```
ls -ld@ "/Volumes/تثبيت-‮١‬"
drwxr-xr-x@ 12 joevt  staff  408 Mar 28 23:34 /Volumes/تثبيت-‮١‬
	com.apple.FinderInfo	 32 


sudo dd if=/dev/rdisk1 skip=2 count=1 2> /dev/null | dd bs=1 skip=0x6c count=4 2> /dev/null | xxd
0000000: 656e 6304                                enc.
```

I wonder if the root directory should have a fusehfs xattr that maps to the script in the MDB?
